apps/speed.c

```diff
@@ -1077,6 +1077,13 @@ static int EdDSA_sign_loop(void *args)
     int ret, count;
 
     for (count = 0; COND(eddsa_c[testnum][0]); count++) {
+        ret = EVP_DigestSignInit(edctx[testnum], NULL, NULL, NULL, NULL);
+        if (ret == 0) {
+            BIO_printf(bio_err, "EdDSA sign init failure\n");
+            ERR_print_errors(bio_err);
+            count = -1;
+            break;
+        }
         ret = EVP_DigestSign(edctx[testnum], eddsasig, eddsasigsize, buf, 20);
```

Comment:
Interestingly the missing EVP_DigestSignInit() was not noticed before - this is really a bug and should be backported to 3.x stable branches.

Reply:
Before commit https://github.com/openssl/openssl/commit/3fc2b7d6b8f961144905330dfd4689f5bd515199 the selected provider's `digest_sign()` function was blindly called, regardless of the state of the context. As long as the provider's  `digest_sign()` function does not check its state, this works. The default provider's EdDSA  `digest_sign()` functions `ed25519_digest_sign()` and `ed448_digest_sign()` don't care about the context state, and are able to produce the correct signature in any case. 

Nevertheless, calling EVP_DigestSign() on a finalized context was always wrong, it just did not hurt with the default provider. For other providers that take more care about the context state it should have always failed.